---
layout: page
title: RNA-Seq data analysis
summary: "RNA-seq data analysis with different approachs."
---

![Pipeline]({{site.url}}/images/pipeline.jpg)

## Introduction

RNA-Seq (RNA sequencing ) also called whole transcriptome sequncing  use next-generation sequeincing (NGS)  to reveal the presence and quantity of RNA in a biolgical sample at a given moment.

<img src="https://upload.wikimedia.org/wikipedia/commons/f/f3/Summary_of_RNA-Seq.svg">

In this tutorial we will try different practical analisys:

- Counting  methods
- Pseudomapping methods



## Practical 1

In this session we want to perform some differential expression from two condition as example (Normal vs tumor RNA-seq). This data use for this tutorial are pubblicaly avaible. We modify that for our purpose.


<img src="{{site.url}}/images/RNA.png" >

Please dwnlad the material here:
   [For count](https://drive.google.com/open?id=1J83O3PhphrskU1Aytke3L7TCIyRm4phQ)
   [Pipeline TEST ](https://drive.google.com/file/d/1ywFJ9OyKn0Bo9AhxX2nUuKC3Fr4wLdL6/view?usp=sharing)
Please follow this  command

    cd ~/Desktop/
    tar xvf TUTORIAL_RNA_snakemake.tar.gz
    cd TUTORIAL
    conda env create -f snakemake_differential.yml
    


Please launch the analysis:


    snakemake -s Snakefile.Kallisto --configfile database3.yml --use-conda

Please Connect and see this tutorial on live sleuth

    [youtube](https://www.youtube.com/watch?v=KEn0CMYk6Wo)


## PRACTICAL COUNTS DATA

In the previous example we use the new way to do RNA-seq analisys using pseudomap approach. This method are recently reported (Kallisto and Salmon).
Another way to do this type analisys  was performed using  counts. Normally you can allign with programs like STAR and count gene and isoforms using HTseq-count or feature-count.
Here we are some examples of working on R on Counts.

<img src="{{site.url}}/images/featurecount.png" >
<img src="{{site.url}}/images/htseq_count.png" >

Please open R-studio


And install  airway package from bioconductor [airway](http://bioconductor.org/packages/release/data/experiment/html/airway.html)
This package provides a RangedSummarizedExperiment object of read counts in genes for an RNA-Seq experiment on four human airway smooth muscle cell lines treated with dexamethasone. Details on the gene model and read counting procedure are provided in the package vignette.

        ### Experimental data
        The data used in this workflow is stored in the airway package that summarizes an RNA-seq experiment wherein airway smooth muscle cells were treated with dexamethasone, a synthetic glucocorticoid steroid with anti-inflammatory effects (Himes et al. 2014). Glucocorticoids are used, for example, by people with asthma to reduce inflammation of the airways. In the experiment, four primary human airway smooth muscle cell lines were treated with 1 micromolar dexamethasone for 18 hours. For each of the four cell lines, we have a treated and an untreated sample. For more description of the experiment see the PubMed entry 24926665 and for raw data see the GEO entry GSE52778.



    source("https://bioconductor.org/biocLite.R")
    biocLite("airway")

Then  activate DESEQ2 and star to read the vignette [DESEq2](http://bioconductor.org/packages/release/bioc/html/DESeq2.html)



      library("airway")
      indir <- system.file("extdata", package="airway", mustWork=TRUE)
      list.files(indir)
We now use Râ€™s data command to load a prepared SummarizedExperiment that was generated from the publicly available sequencing data files associated with Himes et al. (2014), described above. The steps we used to produce this object were equivalent to those you worked through in the previous sections, except that we used all the reads and all the genes. For more details on the exact steps used to create this object, type vignette("airway") into your R session.

            data("airway")
            se <- airway

      ## ------------------------------------------------------------------------
      se$dex %<>% relevel("untrt")
      se$dex

      ## ------------------------------------------------------------------------
      round( colSums(assay(se)) / 1e6, 1 )

      ## ------------------------------------------------------------------------
      colData(se)

      ## ------------------------------------------------------------------------
      library("DESeq2")

      ## ------------------------------------------------------------------------
      dds <- DESeqDataSet(se, design = ~ cell + dex)

      ## ------------------------------------------------------------------------
      countdata <- assay(se)
      head(countdata, 3)

      ## ------------------------------------------------------------------------
      coldata <- colData(se)

      ## ------------------------------------------------------------------------
      ddsMat <- DESeqDataSetFromMatrix(countData = countdata,
                                        colData = coldata,
                                        design = ~ cell + dex)

      ## ------------------------------------------------------------------------
      nrow(dds)
      dds <- dds[ rowSums(counts(dds)) > 1, ]
      nrow(dds)

      ## ----meanSdCts-----------------------------------------------------------
      lambda <- 10^seq(from = -1, to = 2, length = 1000)
      cts <- matrix(rpois(1000*100, lambda), ncol = 100)
      library("vsn")
      meanSdPlot(cts, ranks = FALSE)

      ## ----meanSdLogCts--------------------------------------------------------
      log.cts.one <- log2(cts + 1)
      meanSdPlot(log.cts.one, ranks = FALSE)

      ## ----rlog----------------------------------------------------------------
      rld <- rlog(dds, blind = FALSE)
      head(assay(rld), 3)

      ## ----vst-----------------------------------------------------------------
      vsd <- vst(dds, blind = FALSE)
      head(assay(vsd), 3)

      ## ----rldplot, fig.width = 6, fig.height = 2.5----------------------------
      library("dplyr")
      library("ggplot2")

      dds <- estimateSizeFactors(dds)

      df <- bind_rows(
        as_data_frame(log2(counts(dds, normalized=TRUE)[, 1:2]+1)) %>%
               mutate(transformation = "log2(x + 1)"),
        as_data_frame(assay(rld)[, 1:2]) %>% mutate(transformation = "rlog"),
        as_data_frame(assay(vsd)[, 1:2]) %>% mutate(transformation = "vst"))
        
      colnames(df)[1:2] <- c("x", "y")  

      ggplot(df, aes(x = x, y = y)) + geom_hex(bins = 80) +
        coord_fixed() + facet_grid( . ~ transformation)  

      ## ------------------------------------------------------------------------
      sampleDists <- dist(t(assay(rld)))
      sampleDists

      ## ------------------------------------------------------------------------
      library("pheatmap")
      library("RColorBrewer")

      ## ----distheatmap, fig.width = 6.1, fig.height = 4.5----------------------
      sampleDistMatrix <- as.matrix( sampleDists )
      rownames(sampleDistMatrix) <- paste( rld$dex, rld$cell, sep = " - " )
      colnames(sampleDistMatrix) <- NULL
      colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
      pheatmap(sampleDistMatrix,
               clustering_distance_rows = sampleDists,
               clustering_distance_cols = sampleDists,
               col = colors)

      ## ------------------------------------------------------------------------
      library("PoiClaClu")
      poisd <- PoissonDistance(t(counts(dds)))

      ## ----poisdistheatmap, fig.width = 6.1, fig.height = 4.5------------------
      samplePoisDistMatrix <- as.matrix( poisd$dd )
      rownames(samplePoisDistMatrix) <- paste( rld$dex, rld$cell, sep=" - " )
      colnames(samplePoisDistMatrix) <- NULL
      pheatmap(samplePoisDistMatrix,
               clustering_distance_rows = poisd$dd,
               clustering_distance_cols = poisd$dd,
               col = colors)

      ## ----plotpca, fig.width=6, fig.height=4.5--------------------------------
      plotPCA(rld, intgroup = c("dex", "cell"))

      ## ------------------------------------------------------------------------
      pcaData <- plotPCA(rld, intgroup = c( "dex", "cell"), returnData = TRUE)
      pcaData
      percentVar <- round(100 * attr(pcaData, "percentVar"))

      ## ----ggplotpca, fig.width=6, fig.height=4.5------------------------------
      ggplot(pcaData, aes(x = PC1, y = PC2, color = dex, shape = cell)) +
        geom_point(size =3) +
        xlab(paste0("PC1: ", percentVar[1], "% variance")) +
        ylab(paste0("PC2: ", percentVar[2], "% variance")) +
        coord_fixed()

      ## ----mdsrlog, fig.width=6, fig.height=4.5--------------------------------
      mds <- as.data.frame(colData(rld))  %>%
               cbind(cmdscale(sampleDistMatrix))
      ggplot(mds, aes(x = `1`, y = `2`, color = dex, shape = cell)) +
        geom_point(size = 3) + coord_fixed()

      ## ----mdspois, fig.width=6, fig.height=4.5--------------------------------
      mdsPois <- as.data.frame(colData(dds)) %>%
         cbind(cmdscale(samplePoisDistMatrix))
      ggplot(mdsPois, aes(x = `1`, y = `2`, color = dex, shape = cell)) +
        geom_point(size = 3) + coord_fixed()

      ## ----airwayDE------------------------------------------------------------
      dds <- DESeq(dds)

      ## ------------------------------------------------------------------------
      res <- results(dds)
      res

      ## ------------------------------------------------------------------------
      res <- results(dds, contrast=c("dex","trt","untrt"))

      ## ------------------------------------------------------------------------
      mcols(res, use.names = TRUE)

      ## ------------------------------------------------------------------------
      summary(res)

      ## ------------------------------------------------------------------------
      res.05 <- results(dds, alpha = 0.05)
      table(res.05$padj < 0.05)

      ## ------------------------------------------------------------------------
      resLFC1 <- results(dds, lfcThreshold=1)
      table(resLFC1$padj < 0.1)

      ## ------------------------------------------------------------------------
      results(dds, contrast = c("cell", "N061011", "N61311"))

      ## ----sumres--------------------------------------------------------------
      sum(res$pvalue < 0.05, na.rm=TRUE)
      sum(!is.na(res$pvalue))

      ## ------------------------------------------------------------------------
      sum(res$padj < 0.1, na.rm=TRUE)

      ## ------------------------------------------------------------------------
      resSig <- subset(res, padj < 0.1)
      head(resSig[ order(resSig$log2FoldChange), ])

      ## ------------------------------------------------------------------------
      head(resSig[ order(resSig$log2FoldChange, decreasing = TRUE), ])

      ## ----plotcounts----------------------------------------------------------
      topGene <- rownames(res)[which.min(res$padj)]
      plotCounts(dds, gene = topGene, intgroup=c("dex"))

      ## ----ggplotcountsjitter, fig.width = 4, fig.height = 3-------------------
      library("ggbeeswarm")
      geneCounts <- plotCounts(dds, gene = topGene, intgroup = c("dex","cell"),
                               returnData = TRUE)
      ggplot(geneCounts, aes(x = dex, y = count, color = cell)) +
        scale_y_log10() +  geom_beeswarm(cex = 3)

      ## ----ggplotcountsgroup, fig.width = 4, fig.height = 3--------------------
      ggplot(geneCounts, aes(x = dex, y = count, color = cell, group = cell)) +
        scale_y_log10() + geom_point(size = 3) + geom_line()

      ## ----plotma--------------------------------------------------------------
      res <- lfcShrink(dds, contrast=c("dex","trt","untrt"), res=res)
      plotMA(res, ylim = c(-5, 5))

      ## ----plotmaNoShr---------------------------------------------------------
      res.noshr <- results(dds)
      plotMA(res.noshr, ylim = c(-5, 5))

      ## ----plotmalabel---------------------------------------------------------
      plotMA(res, ylim = c(-5,5))
      topGene <- rownames(res)[which.min(res$padj)]
      with(res[topGene, ], {
        points(baseMean, log2FoldChange, col="dodgerblue", cex=2, lwd=2)
        text(baseMean, log2FoldChange, topGene, pos=2, col="dodgerblue")
      })

      ## ----histpvalue2---------------------------------------------------------
      hist(res$pvalue[res$baseMean > 1], breaks = 0:20/20,
           col = "grey50", border = "white")

      ## ------------------------------------------------------------------------
      library("genefilter")
      topVarGenes <- head(order(rowVars(assay(rld)), decreasing = TRUE), 20)

      ## ----genescluster--------------------------------------------------------
      mat  <- assay(rld)[ topVarGenes, ]
      mat  <- mat - rowMeans(mat)
      anno <- as.data.frame(colData(rld)[, c("cell","dex")])
      pheatmap(mat, annotation_col = anno)

      ## ----sensitivityovermean, fig.width=6------------------------------------
      qs <- c(0, quantile(resLFC1$baseMean[resLFC1$baseMean > 0], 0:6/6))
      bins <- cut(resLFC1$baseMean, qs)
      levels(bins) <- paste0("~", round(signif((qs[-1] + qs[-length(qs)])/2, 2)))
      fractionSig <- tapply(resLFC1$pvalue, bins, function(p)
                                mean(p < .05, na.rm = TRUE))
      barplot(fractionSig, xlab = "mean normalized count",
                           ylab = "fraction of small p values")

      ## ------------------------------------------------------------------------
      library("AnnotationDbi")
      library("org.Hs.eg.db")

      ## ------------------------------------------------------------------------
      columns(org.Hs.eg.db)

      ## ------------------------------------------------------------------------
      res$symbol <- mapIds(org.Hs.eg.db,
                           keys=row.names(res),
                           column="SYMBOL",
                           keytype="ENSEMBL",
                           multiVals="first")
      res$entrez <- mapIds(org.Hs.eg.db,
                           keys=row.names(res),
                           column="ENTREZID",
                           keytype="ENSEMBL",
                           multiVals="first")

      ## ------------------------------------------------------------------------
      resOrdered <- res[order(res$pvalue),]
      head(resOrdered)

      ## ----eval=FALSE----------------------------------------------------------
      ## resOrderedDF <- as.data.frame(resOrdered)[1:100, ]
      ## write.csv(resOrderedDF, file = "results.csv")

     

 well DONE!!!










If you  want to perform all the tutorial plase make ths step :

        source("http://bioconductor.org/workflows.R")
        workflowInstall("rnaseqGene")


and follow this examples [link](https://www.bioconductor.org/help/workflows/rnaseqGene/)






## OPTIONAL PRATICAL

Please follow this tutorial [link] (http://www.nathalievilla.org/doc/html/solution_edgeR-tomato.html#where-to-start-installation-and-alike) Pratical rnaseq data using tomato data






